File: LeetCode/Problems/018_four_sum.rb
# Given an array S of n integers, are there elements a, b, c, and d in S such that a + b + c + d = target? 
# Find all unique quadruplets in the array which gives the sum of target.

# Note: The solution set must not contain duplicate quadruplets.

# For example, given array S = [1, 0, -1, 0, -2, 2], and target = 0.

# A solution set is:
# [
#   [-1,  0, 0, 1],
#   [-2, -1, 1, 2],
#   [-2,  0, 0, 2]
# ]

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
  nums.sort!
  set = []

  (0..nums.length - 3).each do |idx|
    if (idx == 0 || nums[idx] != nums[idx - 1])
      (idx + 1..nums.length - 2).each do |idx2|
        if (idx2 == idx + 1 || nums[idx2] != nums[idx2 - 1])
          low = idx2 + 1
          high = nums.length - 1

          while low < high
            if nums[idx] + nums[idx2] + nums[low] + nums[high] == target
              set << [nums[idx], nums[idx2], nums[low], nums[high]]
              low += 1
              high -= 1

              low += 1 while nums[low] == nums[low - 1]
              high -= 1 while nums[high] == nums[high + 1]
            elsif nums[idx] + nums[idx2] + nums[low] + nums[high] < target
              low += 1
            else
              high -= 1
            end
          end
        end
      end
    end
  end
  
  set
end

File: LeetCode/Problems/018_four_sum.js
const fourSum = (nums, target) => {
  let solutionSet = [];

  nums = nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length - 3; i++) {
    if (i === 0 || nums[i] !== nums[i - 1]) {
      for (let j = i + 1; j < nums.length - 2; j++) {
        if ( j === i + 1 || nums[j] !== nums[j - 1]) {
          let low = j + 1;
          let high = nums.length - 1;

          while (low < high) {
            if (nums[i] + nums[j] + nums[low] + nums[high] === target) {
              solutionSet.push([nums[i], nums[j], nums[low], nums[high]]);
              low++;
              high--;

              while (nums[low] === nums[low - 1]) low++;
              while (nums[high] === nums[high + 1]) high--;
            } else if (nums[i] + nums[j] + nums[low] + nums[high] > target) {
              high--;
            } else {
              low++;
            }
          }
        }
      }
    }
  }

  return solutionSet;
};

let nums = [1, 0, -1, 0, -2, 2];
let target = 0;

console.log(fourSum(nums, target));